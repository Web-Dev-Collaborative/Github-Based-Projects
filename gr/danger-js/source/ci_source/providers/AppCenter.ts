import { Env, CISource } from "../ci_source"
import { ensureEnvKeysExist, getPullRequestIDForBranch } from "../ci_source_helpers"
import * as url from "url"

// AppCenter Build scripts: https://docs.microsoft.com/en-us/appcenter/build/custom/scripts/
// AppCenter Environment variables: https://docs.microsoft.com/en-us/appcenter/build/custom/variables/

/**
 * ### CI Setup
 *
 * To make Danger run, add following lines to the `appcenter-pre-build.sh` file:
 *
 * ```
 * - cd $APPCENTER_SOURCE_DIRECTORY
 * - npm install -g danger
 * - swift build
 * - swift run danger-swift ci
 * ```
 *
 *
 *
 * ### Token Setup
 *
 * Add the `DANGER_GITHUB_API_TOKEN` to your environment variables.
 *
 */

export class AppCenter implements CISource {
  private default = { prID: "0" }
  constructor(private readonly env: Env) {}

  async setup(): Promise<any> {
    const prID = await getPullRequestIDForBranch(this, this.env, this.branchName)
    this.default.prID = prID.toString()
  }

  get name(): string {
    return "AppCenter"
  }

  get isCI(): boolean {
    if (ensureEnvKeysExist(this.env, ["APPCENTER_BUILD_ID"])) {
      return true
    } else {
      return false
    }
  }

  get isPR(): boolean {
    return this.env["BUILD_REASON"] == "PullRequest"
  }

  get pullRequestID(): string {
    return this.default.prID
  }

  get repoSlug(): string {
    if (
      ensureEnvKeysExist(this.env, ["BUILD_REPOSITORY_NAME"]) &&
      ensureEnvKeysExist(this.env, ["BUILD_REPOSITORY_URI"])
    ) {
      const repositoryName = this.env["BUILD_REPOSITORY_NAME"]
      const components = url.parse(this.env["BUILD_REPOSITORY_URI"], false)
      if (components && components.path) {
        const owner = components.path.split("/")[1]
        return `${owner}/${repositoryName}`
      }
    }
    return ""
  }

  private get branchName(): string {
    return this.env["BUILD_SOURCEBRANCHNAME"]
  }
}
