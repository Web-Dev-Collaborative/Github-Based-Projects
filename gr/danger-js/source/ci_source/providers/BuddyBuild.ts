import { Env, CISource } from "../ci_source"
import { ensureEnvKeysExist, ensureEnvKeysAreInt } from "../ci_source_helpers"

/**
 * ### CI Setup
 *
 * Buddybuild has an integration for Danger JS already built-in.
 *
 * ### Token Setup
 *
 * Login to buddybuild and select your app. Go to your *App Settings* and
 * in the *Build Settings* menu on the left, choose *Environment Variables*.
 *
 * #### GitHub
 *
 * Add the `DANGER_GITHUB_API_TOKEN` to your build user's ENV.
 *
 */
export class BuddyBuild implements CISource {
  constructor(private readonly env: Env) {}

  get name(): string {
    return "buddybuild"
  }

  get isCI(): boolean {
    return ensureEnvKeysExist(this.env, ["BUDDYBUILD_BUILD_ID"])
  }

  get isPR(): boolean {
    const mustHave = ["BUDDYBUILD_PULL_REQUEST", "BUDDYBUILD_REPO_SLUG"]
    const mustBeInts = ["BUDDYBUILD_PULL_REQUEST"]
    return ensureEnvKeysExist(this.env, mustHave) && ensureEnvKeysAreInt(this.env, mustBeInts)
  }

  get pullRequestID(): string {
    return this.env.BUDDYBUILD_PULL_REQUEST
  }

  get repoSlug(): string {
    return this.env.BUDDYBUILD_REPO_SLUG
  }
}
