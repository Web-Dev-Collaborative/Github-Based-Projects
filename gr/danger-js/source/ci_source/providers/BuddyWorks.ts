import { Env, CISource } from "../ci_source"
import { ensureEnvKeysExist, ensureEnvKeysAreInt } from "../ci_source_helpers"
/**
 * ### CI Setup
 *
 * To make Danger run:
 *
 * - Create a new pipeline named, DangerJS, which is triggered on every push
 * - Add a NodeJS environment as an Action
 * - Go into it, head over to the bash editor and type the following
 *    - `yarn install && yarn danger ci`
 *    - or your npm script
 * - Set the `DANGER_GITHUB_API_TOKEN` at the Variables section
 * - You're done ðŸŽ‰
 *
 */
export class BuddyWorks implements CISource {
  constructor(private readonly env: Env) {}

  get name(): string {
    return "Buddy.works"
  }

  get isCI(): boolean {
    return ensureEnvKeysExist(this.env, ["BUDDY_PIPELINE_ID"])
  }

  get isPR(): boolean {
    const mustHave = ["BUDDY_PIPELINE_ID", "BUDDY_EXECUTION_PULL_REQUEST_NO", "BUDDY_REPO_SLUG"]
    const mustBeInts = ["BUDDY_EXECUTION_PULL_REQUEST_NO"]

    return ensureEnvKeysExist(this.env, mustHave) && ensureEnvKeysAreInt(this.env, mustBeInts)
  }

  get pullRequestID(): string {
    return this.env.BUDDY_EXECUTION_PULL_REQUEST_NO
  }

  get repoSlug(): string {
    return this.env.BUDDY_REPO_SLUG
  }

  get ciRunURL() {
    return this.env.BUDDY_EXECUTION_URL
  }
}

// Default ENV vars provided by Buddy.works
// https://buddy.works/docs/pipelines/environment-variables#default-environment-variables
