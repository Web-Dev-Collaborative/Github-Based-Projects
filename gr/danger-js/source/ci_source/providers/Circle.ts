import { Env, CISource } from "../ci_source"
import { ensureEnvKeysExist, ensureEnvKeysAreInt } from "../ci_source_helpers"

/**
 *  ### CI Setup
 *
 *  For setting up Circle CI, we recommend turning on "Only Build pull requests." in "Advanced Setting." Without this enabled,
 *  it is _really_ tricky for Danger to know whether you are in a pull request or not, as the environment metadata
 *  isn't reliable.
 *
 * <!-- JS --!>
 *  With that set up, you can you add `yarn danger ci` to your `circle.yml`. If you override the default
 *  `test:` section, then add it as an extra step to the list.
 *
 *  ```yml
 *   - run:
 *       name: Danger
 *       command: yarn danger ci
 *  ```
 * <!-- !JS --!>
 * <!-- Swift --!>
 *
 *  Add some build steps to make Danger Swift work:
 *
 *  ```yml
 *   - run:
 *       name: Installing Danger JS
 *       command: npm install -g danger
 *   - run:
 *       name: Compiling Danger
 *       command: swift build
 *   - run:
 *       name: Running Danger Swift
 *       command: swift run danger-swift ci
 *  ```
 *
 * We'd also recommend adding both `.build` and `~/.danger-swift` to your build cache too.
 *
 * <!-- !Swift --!>
 *
 *  ### Token Setup
 *
 *  There is no difference here for OSS vs Closed, add your `DANGER_GITHUB_API_TOKEN` to the Environment variable settings page.
 *
 */
export class Circle implements CISource {
  constructor(private readonly env: Env) {}

  get name(): string {
    return "Circle CI"
  }

  get isCI(): boolean {
    return ensureEnvKeysExist(this.env, ["CIRCLE_BUILD_NUM"])
  }

  get isPR(): boolean {
    if (ensureEnvKeysExist(this.env, ["CI_PULL_REQUEST"]) || ensureEnvKeysExist(this.env, ["CIRCLE_PULL_REQUEST"])) {
      return true
    }

    const mustHave = ["CIRCLE_CI_API_TOKEN", "CIRCLE_PROJECT_USERNAME", "CIRCLE_PROJECT_REPONAME", "CIRCLE_BUILD_NUM"]
    return ensureEnvKeysExist(this.env, mustHave) && ensureEnvKeysAreInt(this.env, ["CIRCLE_PR_NUMBER"])
  }

  private _prParseURL(): { owner?: string; reponame?: string; id?: string } {
    const prUrl = this.env.CI_PULL_REQUEST || this.env.CIRCLE_PULL_REQUEST || ""
    const splitSlug = prUrl.split("/")
    if (splitSlug.length === 7) {
      const owner = splitSlug[3]
      const reponame = splitSlug[4]
      const id = splitSlug[6]
      return { owner, reponame, id }
    }
    return {}
  }

  get pullRequestID(): string {
    if (this.env.CIRCLE_PR_NUMBER) {
      return this.env.CIRCLE_PR_NUMBER
    } else {
      const { id } = this._prParseURL()
      return id || ""
    }
  }

  get repoSlug(): string {
    const { owner, reponame } = this._prParseURL()
    return owner && reponame ? `${owner}/${reponame}` : ""
  }

  get repoURL(): string {
    return this.env.CIRCLE_REPOSITORY_URL
  }

  get ciRunURL() {
    return this.env["CIRCLE_BUILD_URL"]
  }
}
