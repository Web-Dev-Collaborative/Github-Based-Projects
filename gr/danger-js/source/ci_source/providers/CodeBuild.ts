import { Env, CISource } from "../ci_source"
import { ensureEnvKeysExist } from "../ci_source_helpers"

/**
 * CI Setup
 *
 * In CodeBuild, make sure to correctly forward CODEBUILD_BUILD_ID, CODEBUILD_SOURCE_VERSION, CODEBUILD_SOURCE_REPO_URL and DANGER_GITHUB_API_TOKEN.
 *
 * Token Setup
 *
 * Add your `DANGER_GITHUB_API_TOKEN` to your project. Edit -> Environment -> Additional configuration -> Create a parameter
 *
 */
export class CodeBuild implements CISource {
  constructor(private readonly env: Env) {}

  get name(): string {
    return "CodeBuild"
  }

  get isCI(): boolean {
    return ensureEnvKeysExist(this.env, ["CODEBUILD_BUILD_ID"])
  }

  get isPR(): boolean {
    const mustHave = ["CODEBUILD_BUILD_ID", "CODEBUILD_SOURCE_VERSION", "CODEBUILD_SOURCE_REPO_URL"]
    return ensureEnvKeysExist(this.env, mustHave) && this._isPRRequest()
  }

  get pullRequestID(): string {
    return this.env.CODEBUILD_SOURCE_VERSION.split("/")[1]
  }

  get repoSlug(): string {
    return this._prParseUrl()
  }

  get repoURL(): string {
    return this.env.CODEBUILD_SOURCE_REPO_URL
  }

  private _isPRRequest(): boolean {
    const isPRSource = this.env.CODEBUILD_SOURCE_VERSION.split("/")[0] === "pr" ? true : false
    const isPRIdInt = !isNaN(parseInt(this.env.CODEBUILD_SOURCE_VERSION.split("/")[1]))
    return isPRSource && isPRIdInt
  }

  private _prParseUrl(): string {
    const prUrl = this.env.CODEBUILD_SOURCE_REPO_URL || ""
    const regexp = new RegExp("([/:])([^/]+/[^/.]+)(?:.git)?$")
    const matches = prUrl.match(regexp)
    return matches ? matches[2] : ""
  }
}
