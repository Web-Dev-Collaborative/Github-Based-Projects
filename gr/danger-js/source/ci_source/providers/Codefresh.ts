import { Env, CISource } from "../ci_source"
import { ensureEnvKeysExist, ensureEnvKeysAreInt } from "../ci_source_helpers"

/**
 *  ### CI Setup
 *
 *  To set up Danger on Codefresh, create a freestyle step in your Codefresh yaml configuration:
 *
 *  ```yml
 *  Danger:
 *    title: Run Danger
 *    image: node:latest
 *    working_directory: ${{main_clone}}
 *    entry_point: '/bin/bash'
 *    cmd:
 *      - '-ce'
 *      - |
 *        npm install -g yarn
 *        yarn add danger --dev
 *        yarn danger ci --failOnErrors
 *    when:
 *      steps:
 *        - name: main_clone
 *          on:
 *            - success
 *  ```
 *
 *  The `failOnErrors` option is required in order to ensure that the step fails properly when Danger fails. If you don't wnat this behavior, you can remove this option.
 *
 *  Don't forget to add the `DANGER_GITHUB_API_TOKEN` variable to your pipeline settings so that Danger can properly post comments to your pull request.
 *
 */

export class Codefresh implements CISource {
  constructor(private readonly env: Env) {}

  get name(): string {
    return "Codefresh"
  }

  get isCI(): boolean {
    return ensureEnvKeysExist(this.env, ["CF_BUILD_ID", "CF_BUILD_URL"])
  }

  get isPR(): boolean {
    return (
      ensureEnvKeysExist(this.env, ["CF_PULL_REQUEST_NUMBER"]) &&
      ensureEnvKeysAreInt(this.env, ["CF_PULL_REQUEST_NUMBER"])
    )
  }

  get pullRequestID(): string {
    if (this.env.CF_PULL_REQUEST_NUMBER) {
      return this.env.CF_PULL_REQUEST_NUMBER
    } else {
      return ""
    }
  }

  get repoSlug(): string {
    const splitSlug = this.env.CF_COMMIT_URL.split("/")
    if (splitSlug.length === 7) {
      const owner = splitSlug[3]
      const reponame = splitSlug[4]
      return owner && reponame ? `${owner}/${reponame}` : ""
    }
    return ""
  }

  get ciRunURL() {
    return this.env.CF_BUILD_URL
  }
}
