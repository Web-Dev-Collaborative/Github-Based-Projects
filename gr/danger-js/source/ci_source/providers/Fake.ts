import { Env, CISource } from "../ci_source"
import { ensureEnvKeysExist } from "../ci_source_helpers"

export class FakeCI implements CISource {
  private readonly env: Env

  constructor(env: Env) {
    const defaults = {
      repo: env.DANGER_TEST_REPO || "artsy/emission", // TODO: default to empty string ?
      pr: env.DANGER_TEST_PR || "327", // TODO: default to empty string ?
    }

    this.env = { ...env, ...defaults }
  }
  get name(): string {
    return "Fake Testing CI"
  }

  get isCI(): boolean {
    return ensureEnvKeysExist(this.env, ["DANGER_FAKE_CI"]) || ensureEnvKeysExist(this.env, ["DANGER_LOCAL_NO_CI"])
  }

  get isPR(): boolean {
    return true
  }

  get pullRequestID(): string {
    return this.env.pr
  }

  get repoSlug(): string {
    return this.env.repo
  }
}
