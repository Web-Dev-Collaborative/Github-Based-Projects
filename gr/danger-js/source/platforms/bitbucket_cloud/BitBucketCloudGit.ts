import { GitJSONDSL, GitDSL } from "../../dsl/GitDSL"
import { BitBucketCloudAPI } from "./BitBucketCloudAPI"
import { diffToGitJSONDSL } from "../git/diffToGitJSONDSL"
import { BitBucketCloudCommit, BitBucketCloudDSL } from "../../dsl/BitBucketCloudDSL"
import { GitCommit, GitCommitAuthor } from "../../dsl/Commit"
import { GitJSONToGitDSLConfig, gitJSONToGitDSL } from "../git/gitJSONToGitDSL"

import { debug } from "../../debug"
const d = debug("BitBucketCloudGit")

export function bitBucketCloudRawAndDateToGitCommitAuthor(raw: string, date: string): GitCommitAuthor {
  const startIndexOfEmail = raw.lastIndexOf("<")
  const endIndexOfEmail = raw.lastIndexOf(">")
  if (startIndexOfEmail === -1 || endIndexOfEmail === -1) {
    return {
      name: raw,
      email: "",
      date,
    }
  }
  const name = raw.substring(0, startIndexOfEmail).trim()
  const email = raw.substring(startIndexOfEmail + 1, endIndexOfEmail).trim()
  return {
    name,
    email,
    date,
  }
}

function bitBucketCloudCommitToGitCommit(commit: BitBucketCloudCommit): GitCommit {
  const user = bitBucketCloudRawAndDateToGitCommitAuthor(commit.author.raw, commit.date)

  return {
    sha: commit.hash,
    author: user,
    committer: user,
    message: commit.message,
    tree: null,
    parents: commit.parents != null ? commit.parents.map(parent => parent.hash) : undefined,
    url: commit.links.html.href,
  }
}

export default async function gitDSLForBitBucketCloud(api: BitBucketCloudAPI): Promise<GitJSONDSL> {
  // We'll need all this info to be able to generate a working GitDSL object
  const diffs = await api.getPullRequestDiff()
  const gitCommits = await api.getPullRequestCommits()
  const commits = gitCommits.map(bitBucketCloudCommitToGitCommit)
  return diffToGitJSONDSL(diffs, commits)
}

export const bitBucketCloudGitDSL = (
  bitBucketCloud: BitBucketCloudDSL,
  json: GitJSONDSL,
  bitBucketCloudAPI: BitBucketCloudAPI
): GitDSL => {
  const config: GitJSONToGitDSLConfig = {
    repo: bitBucketCloud.pr.source.repository.full_name,
    baseSHA: bitBucketCloud.pr.destination.commit.hash,
    headSHA: bitBucketCloud.pr.source.commit.hash,
    getFileContents: bitBucketCloudAPI.getFileContents,
    getFullDiff: bitBucketCloudAPI.getPullRequestDiff,
  }

  d("Setting up git DSL with: ", config)
  return gitJSONToGitDSL(json, config)
}
