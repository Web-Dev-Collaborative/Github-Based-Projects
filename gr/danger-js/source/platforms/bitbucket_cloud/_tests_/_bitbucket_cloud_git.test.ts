import { bitBucketCloudRawAndDateToGitCommitAuthor } from "../BitBucketCloudGit"

import { BitBucketCloud } from "../../BitBucketCloud"
import { BitBucketCloudAPI } from "../BitBucketCloudAPI"
import { FakeCI } from "../../../ci_source/providers/Fake"
import { readFileSync, writeFileSync } from "fs"
import { resolve, join as pathJoin } from "path"
import { bitBucketCloudGitDSL as gitJSONToGitDSL } from "../BitBucketCloudGit"

import { BitBucketCloudDSL } from "../../../dsl/BitBucketCloudDSL"
import { GitJSONDSL, GitDSL } from "../../../dsl/GitDSL"
import { GitCommit } from "../../../dsl/Commit"
import { jsonDSLGenerator } from "../../../runner/dslGenerator"

const fixtures = resolve(__dirname, "..", "..", "_tests", "fixtures")

/** Returns a fixture. */
const loadFixture = (path: string): string =>
  readFileSync(pathJoin(fixtures, path), {})
    .toString()
    .replace(/\r/g, "")

/** Returns JSON from the fixtured dir */
export const requestWithFixturedJSON = async (path: string): Promise<() => Promise<any>> => () =>
  Promise.resolve(JSON.parse(loadFixture(path)))

/** Returns arbitrary text value from a request */
export const requestWithFixturedContent = async (path: string): Promise<() => Promise<string>> => () =>
  Promise.resolve(loadFixture(path))

/**
 * HACKish: Jest on Windows seems to include some additional
 * whitespace that differs from non-Windows snapshot output,
 * so we strip these until we can better-address the issue.
 */
const stripWhitespaceForSnapshot = (str: string) => {
  return str.replace(/\s/g, "")
}

const pullRequestInfoFilename = "bitbucket_cloud_pr.json"

describe("the dangerfile gitDSL - BitBucket Cloud", () => {
  let bbc: BitBucketCloud = {} as any
  let gitJSONDSL: GitJSONDSL = {} as any

  let bbcDSL: BitBucketCloudDSL = {} as any
  let gitDSL: GitDSL = {} as any

  beforeEach(async () => {
    const api = new BitBucketCloudAPI(new FakeCI({}), {
      username: "username",
      password: "password",
      uuid: "{1234-1234-1234-1234}",
      type: "PASSWORD",
    })
    bbc = new BitBucketCloud(api)

    api.getPullRequestInfo = await requestWithFixturedJSON(pullRequestInfoFilename)
    api.getPullRequestComments = await requestWithFixturedJSON("bitbucket_cloud_comments.json")
    api.getPullRequestCommits = await requestWithFixturedJSON("bitbucket_cloud_commits.json")
    api.getPullRequestActivities = await requestWithFixturedJSON("bitbucket_cloud_activities.json")
    api.getPullRequestDiff = await requestWithFixturedContent("bitbucket_cloud_diff.diff")
    api.getFileContents = async (path, repoSlug, ref) => JSON.stringify({ path, repoSlug, ref })

    gitJSONDSL = await bbc.getPlatformGitRepresentation()
    bbcDSL = await bbc.getPlatformReviewDSLRepresentation()
    gitDSL = gitJSONToGitDSL(bbcDSL, gitJSONDSL, bbc.api)
  })

  it("sets the modified/created/deleted", async () => {
    expect(gitJSONDSL.modified_files).toEqual([
      "CHANGELOG.md",
      "data/schema.graphql",
      "data/schema.json",
      "externals/metaphysics",
      "lib/__mocks__/react-relay.js",
      "lib/components/artist/about.js",
      "lib/components/gene/header.js",
      "lib/containers/__tests__/__snapshots__/gene-tests.js.snap",
      "lib/containers/__tests__/gene-tests.js",
      "lib/containers/gene.js",
      "tsconfig.json",
    ])

    expect(gitJSONDSL.created_files).toEqual([
      "lib/components/gene/about.js",
      "lib/components/gene/biography.js",
      "lib/components/related_artists/index.js",
      "lib/components/related_artists/related_artist.js",
    ])

    expect(gitJSONDSL.deleted_files).toEqual([
      "lib/components/artist/related_artists/index.js",
      "lib/components/artist/related_artists/related_artist.js",
      "lib/components/gene/about_gene.js",
    ])
  })

  it("shows the diff for a specific file", async () => {
    const { diff } = (await gitDSL.diffForFile("tsconfig.json"))!

    expect(stripWhitespaceForSnapshot(diff)).toMatchSnapshot()
  })

  it("should include `before` text content of the file", async () => {
    const { before } = (await gitDSL.diffForFile("tsconfig.json"))!

    expect(stripWhitespaceForSnapshot(before)).toMatchSnapshot()
  })

  it("should include `after` text content of the file", async () => {
    const { after } = (await gitDSL.diffForFile("tsconfig.json"))!

    expect(stripWhitespaceForSnapshot(after)).toMatchSnapshot()
  })

  it("should include `added` text content of the file", async () => {
    const { added } = (await gitDSL.diffForFile("tsconfig.json"))!

    expect(stripWhitespaceForSnapshot(added)).toMatchSnapshot()
  })

  it("should include `removed` text content of the file", async () => {
    const { removed } = (await gitDSL.diffForFile("tsconfig.json"))!

    expect(stripWhitespaceForSnapshot(removed)).toMatchSnapshot()
  })

  it("resolves to `null` for files not in modified_files", async () => {
    const result = (await gitDSL.diffForFile("fuhqmahgads.json"))!

    expect(result).toBeNull()
  })

  it("shows the structured diff for a specific file", async () => {
    const { chunks } = (await gitDSL.structuredDiffForFile("tsconfig.json"))!
    expect(chunks).toMatchSnapshot()
  })

  it("sets up commit data correctly", async () => {
    const exampleCommit: GitCommit = {
      author: {
        date: "2019-05-13T13:24:53+00:00",
        email: "foo@bar.com",
        name: "Hello Core",
      },
      committer: {
        date: "2019-05-13T13:24:53+00:00",
        email: "foo@bar.com",
        name: "Hello Core",
      },
      message: "Merged in develop (pull request #174)\n\nDevelop to SIT\n\nApproved-by: Hello Core <foo@bar.com>\n",
      parents: ["77fdf99fcc08dce8bd49c9aedfc7442f2bd40bde", "a584f1256b8522cfce4ccc33dfbf157590a6044d"],
      sha: "7f73207cea935b2fd07a76028484c20eb18422ba",
      tree: null,
      url: "https://bitbucket.org/foo/bar/commits/7f73207cea935b2fd07a76028484c20eb18422ba",
    }

    expect(gitDSL.commits[0]).toEqual(exampleCommit)
  })

  it("writes a JSON DSL fixture", async () => {
    const fakeSource = new FakeCI({})
    const dataSent = await jsonDSLGenerator(bbc, fakeSource, {} as any)
    dataSent.settings.github.accessToken = "12345"

    writeFileSync(pathJoin(fixtures, "bbc-dsl-input.json"), JSON.stringify(dataSent, null, "  "), "utf8")
  })
})

describe("bitBucketCloudRawAndDateToGitCommitAuthor", () => {
  const date = "2019-05-13T11:41:13+00:00"
  it("should convert name doesn't contain space correctly", () => {
    const raw = "Foo <foo@bar.com>"
    expect(bitBucketCloudRawAndDateToGitCommitAuthor(raw, date)).toEqual({
      name: "Foo",
      email: "foo@bar.com",
      date,
    })
  })
  it("should convert name contains one space correctly", () => {
    const raw = "Foo Bar <foo@bar.com>"
    expect(bitBucketCloudRawAndDateToGitCommitAuthor(raw, date)).toEqual({
      name: "Foo Bar",
      email: "foo@bar.com",
      date,
    })
  })
  it("should convert name contains multiple space correctly", () => {
    const raw = "Foo Bar Foo Bar Foo <foo@bar.com>"
    expect(bitBucketCloudRawAndDateToGitCommitAuthor(raw, date)).toEqual({
      name: "Foo Bar Foo Bar Foo",
      email: "foo@bar.com",
      date,
    })
  })
  it("should convert name contains special characters correctly", () => {
    const raw = "Foo Bar < Foo  @Bar >Foo <foo@bar.com>"
    expect(bitBucketCloudRawAndDateToGitCommitAuthor(raw, date)).toEqual({
      name: "Foo Bar < Foo  @Bar >Foo",
      email: "foo@bar.com",
      date,
    })
  })
  it("should convert email contains multiple dot correctly", () => {
    const raw = "Foo Bar <foo@bar.hello.com>"
    expect(bitBucketCloudRawAndDateToGitCommitAuthor(raw, date)).toEqual({
      name: "Foo Bar",
      email: "foo@bar.hello.com",
      date,
    })
  })
  it("should put raw into name if it couldn't convert", () => {
    const raw = "Foo Bar"
    expect(bitBucketCloudRawAndDateToGitCommitAuthor(raw, date)).toEqual({
      name: raw,
      email: "",
      date,
    })
  })
  it("should put only name if it couldn't find an email", () => {
    const raw = "Foo Bar <>"
    expect(bitBucketCloudRawAndDateToGitCommitAuthor(raw, date)).toEqual({
      name: "Foo Bar",
      email: "",
      date,
    })
  })
})
