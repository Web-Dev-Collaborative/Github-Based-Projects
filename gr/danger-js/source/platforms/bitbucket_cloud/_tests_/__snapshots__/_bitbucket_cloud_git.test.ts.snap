// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the dangerfile gitDSL - BitBucket Cloud should include \`added\` text content of the file 1`] = `"+\\"Example/Build\\",+\\"externals/\\""`;

exports[`the dangerfile gitDSL - BitBucket Cloud should include \`after\` text content of the file 1`] = `"{\\"path\\":\\"tsconfig.json\\",\\"repoSlug\\":\\"foo/bar\\",\\"ref\\":\\"007bf2423436\\"}"`;

exports[`the dangerfile gitDSL - BitBucket Cloud should include \`before\` text content of the file 1`] = `"{\\"path\\":\\"tsconfig.json\\",\\"repoSlug\\":\\"foo/bar\\",\\"ref\\":\\"8a2eb414cb5d\\"}"`;

exports[`the dangerfile gitDSL - BitBucket Cloud should include \`removed\` text content of the file 1`] = `"-\\"Example/Build\\""`;

exports[`the dangerfile gitDSL - BitBucket Cloud shows the diff for a specific file 1`] = `"\\"exclude\\":[\\"node_modules\\",\\"Pod/Assets\\",-\\"Example/Build\\"+\\"Example/Build\\",+\\"externals/\\"]}"`;

exports[`the dangerfile gitDSL - BitBucket Cloud shows the structured diff for a specific file 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "     \\"exclude\\": [",
        "ln1": 5,
        "ln2": 5,
        "normal": true,
        "type": "normal",
      },
      Object {
        "content": "         \\"node_modules\\",",
        "ln1": 6,
        "ln2": 6,
        "normal": true,
        "type": "normal",
      },
      Object {
        "content": "         \\"Pod/Assets\\",",
        "ln1": 7,
        "ln2": 7,
        "normal": true,
        "type": "normal",
      },
      Object {
        "content": "-        \\"Example/Build\\"",
        "del": true,
        "ln": 8,
        "type": "del",
      },
      Object {
        "add": true,
        "content": "+        \\"Example/Build\\",",
        "ln": 8,
        "type": "add",
      },
      Object {
        "add": true,
        "content": "+        \\"externals/\\"",
        "ln": 9,
        "type": "add",
      },
      Object {
        "content": "     ]",
        "ln1": 9,
        "ln2": 10,
        "normal": true,
        "type": "normal",
      },
      Object {
        "content": " }",
        "ln1": 10,
        "ln2": 11,
        "normal": true,
        "type": "normal",
      },
    ],
    "content": "@@ -5,6 +5,7 @@",
    "newLines": 7,
    "newStart": 5,
    "oldLines": 6,
    "oldStart": 5,
  },
]
`;
